From e06ef7a848eeadd1ee1969997e304524356b1805 Mon Sep 17 00:00:00 2001
From: Andrew Leonard <anleonar@redhat.com>
Date: Thu, 23 Feb 2023 11:17:22 +0000
Subject: [PATCH] Regen generated-configure.sh as contains conflict msgs

Signed-off-by: Andrew Leonard <anleonar@redhat.com>
---
 common/autoconf/generated-configure.sh | 10 +---------
 1 file changed, 1 insertion(+), 9 deletions(-)

diff --git a/common/autoconf/generated-configure.sh b/common/autoconf/generated-configure.sh
index a08bec3219..00a210900a 100644
--- a/common/autoconf/generated-configure.sh
+++ b/common/autoconf/generated-configure.sh
@@ -4994,15 +4994,7 @@ VS_TOOLSET_SUPPORTED_2019=false
 #CUSTOM_AUTOCONF_INCLUDE
 
 # Do not change or remove the following line, it is needed for consistency checks:
-<<<<<<< HEAD
-<<<<<<< HEAD
-DATE_WHEN_GENERATED=1673650816
-=======
-DATE_WHEN_GENERATED=1670219878
->>>>>>> origin/master
-=======
-DATE_WHEN_GENERATED=1676455289
->>>>>>> origin/release
+DATE_WHEN_GENERATED=1677150961
 
 ###############################################################################
 #
-- 
2.52.0

