From 748e72a2784e6062dd4fe2cc624ddf0e4445fe73 Mon Sep 17 00:00:00 2001
From: Andrew Leonard <anleonar@redhat.com>
Date: Thu, 23 Feb 2023 11:18:52 +0000
Subject: [PATCH] Regen generated-configure.sh as contains conflict msgs

Signed-off-by: Andrew Leonard <anleonar@redhat.com>
---
 common/autoconf/generated-configure.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/common/autoconf/generated-configure.sh b/common/autoconf/generated-configure.sh
index e12e97757f..358b389ad0 100644
--- a/common/autoconf/generated-configure.sh
+++ b/common/autoconf/generated-configure.sh
@@ -4994,7 +4994,7 @@ VS_TOOLSET_SUPPORTED_2019=false
 #CUSTOM_AUTOCONF_INCLUDE
 
 # Do not change or remove the following line, it is needed for consistency checks:
-DATE_WHEN_GENERATED=1676455289
+DATE_WHEN_GENERATED=1677151107
 
 ###############################################################################
 #
-- 
2.52.0

